### ISSUE TITLE
tilemap.loadFrom2DArray makes map data immutable

### ISSUE THREAD

**Issue by matrefeytontias** [HaxePunk/HaxePunk#358]
It is worth noting that in [the function's definition in Tilemap.hx](https://github.com/HaxePunk/HaxePunk/blob/master/com/haxepunk/graphics/Tilemap.hx#L241-L254), removing the `if` condition (not what's inside the block of course) and the `_map = array;` line makes the map data mutable again.

**Comment by Marc010** [comment]
Try to change the code to this:
```
	public function loadFrom2DArray(array:Array<Array<Int>>):Void
	{
		if (blit)
		{
			for (y in 0...array.length)
			 {
				for (x in 0...array[0].length)
				{
					setTile(x, y, array[y][x]);
				}
			 }
		}
		else
		{
			_map = array;
		}
	}
```
It would be nice if you could test if this works fine on flash and neko/windows.

**Comment by matrefeytontias** [comment]
I did try that and I didn't work, because apparently blit is false (or null ?) by default, so _map gets set to array anyway. All the testing I did was on native Windows, I'll try with neko and flash right now.

**Comment by matrefeytontias** [comment]
Mh, for some reason I can't test my project with Neko ; even without the modification to Tilemap.hx it builds fine but crashes with `Uncaught exception - std@module_read`, and every call to `trace()` I can do make no sign of existence when the program is ran, even when put in main() right before the Main class creation. I'll try testing with flash while I sort that up.

**Comment by Marc010** [comment]
Actually there's a good chance that just removing the if and _map assignment will work  because because the blit related stuff was already added to setTile().

**Comment by scriptorum** [comment]
I can see there being issues if you don't copy() the array before passing it in, but I doubt that will account for what you're describing.

**Comment by scriptorum** [comment]
But yeah Marc's right, there should be an else in there.

**Comment by matrefeytontias** [comment]
As I said 5 comments ago, I tried and it doesn't work because `blit` seems to be false or null by default. The problem is actually the line `_map = array;`, and if like Marc010 says it, this thing is taken care of somewhere else, there's no need to find a workaround when you can just remove it, right ?

**Comment by ibilon** [comment]
Yeah the blit related code is guarded in setTile so it's not really a problem to remove it, and it's not like loadFrom2Darray is called often so there's not really a cost in copying the array, and if it solve some problems.
Though I wonder what causes the bug.

**Comment by Marc010** [comment]
loadfrom2DArray() calls setTile() for every tile so
```_map[row][column] = index;``` from setTile() makes it redundant to assign the _map var

**Comment by ibilon** [comment]
Yeah we'd need to remove the assignment, that would also prevent the user to modify the size of that array.

**Comment by scriptorum** [comment]
I put together my own Tilemap test and while there are some issues on Flash when using a transparent tileset (working on a fix for that), I had no problems animating using setTile or loadFrom2DArray on native. Hrm.

**Comment by matrefeytontias** [comment]
Aaaaaaaalright, I have no idea what happened, but I randomly had the idea to update hxcpp and that actually fixed it, now the tilemap is being updated correctly, even without any change to Tilemap.hx. Welp, I guess that still helped to find redundant code.

**Comment by matrefeytontias** [comment]
Last time I had it updated was 2 months ago I believe.

**Comment by ibilon** [comment]
Maybe some issues with memory management and the use of pointer vs object copy, well at least your problem is resolved :)

**Comment by scriptorum** [comment]
@matrefeytontias Good to hear. 

@ibilon Since I'm already working Tilemap, I'll remove the map assignment in the loadXXX calls and have everything go through setTile.

**Comment by matrefeytontias** [comment]
Haha alright never mind, that was because I changed some code.

Alright so that's weird. I have this:
```Haxe
var file:FileInput = File.read("data/maps/" + level + "/tilemap.json");
var tilemapData:Dynamic = haxe.Json.parse(file.readAll().toString());
file.close();
```
Then there's tilemap.json, which is this:
```
{
	"map": [
			[ -1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, -1 ],
			[ 1, 20, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 15, 2 ],
			[ 1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2 ],
			[ 1, 4, 4, 20, 15, 15, 15, 15, 15, 15, 15, 15, 15, 4, 4, 2 ],
			[ 1, 4, 4, 20, 4, 4, 4, 4, 4, 4, 4, 4, 5, 4, 4, 2 ],
			[ 1, 4, 4, 20, 4, 4, 4, 4, 4, 4, 4, 4, 5, 4, 4, 2 ],
			[ 1, 4, 4, 20, 4, 4, 4, 4, 4, 4, 4, 4, 5, 4, 4, 2 ],
			[ 1, 4, 4, 20, 4, 4, 4, 4, 4, 4, 4, 4, 5, 4, 4, 2 ],
			[ 1, 4, 4, 20, 4, 4, 4, 4, 4, 4, 4, 4, 5, 4, 4, 2 ],
			[ 1, 4, 4, 20, 4, 4, 4, 4, 4, 4, 4, 4, 5, 4, 4, 2 ],
			[ 1, 4, 4, 20, 4, 4, 4, 4, 4, 4, 4, 4, 5, 4, 4, 2 ],
			[ 1, 4, 4, 20, 4, 4, 4, 4, 4, 4, 4, 4, 5, 4, 4, 2 ],
			[ 1, 4, 4, 10, 10, 10, 10, 10, 10, 10, 10, 10, 5, 4, 4, 2 ],
			[ 1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2 ],
			[ 1, 10, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 2 ],
			[ -1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1 ] ],
	"rects": [
			[ 0, 0, 16, 1 ],
			[ 15, 1, 1, 15 ],
			[ 0, 1, 1, 15 ],
			[ 1, 15, 14, 1 ]
			]
}
```
So when I just create `tilemapData:Dynamic` and assign it this exact data from the JSON file, it works great when I do `tiles.loadFrom2DArray(tilemapData.map);`. But when I do the code above, it breaks. Am I missing an important Haxe feature or is that a bug ?

**Comment by matrefeytontias** [comment]
It looks like it is the `Dynamic` given by `haxe.Json.parse` that is immutable, and remains immutable when assigned to _map via `_map = array;`. I have no idea if this is the expected behaviour.

**Comment by matrefeytontias** [comment]
And by "it breaks", I mean the tilemap itself is fine, but it won't animate.

**Comment by scriptorum** [comment]
If you do tiles.loadFrom2DArray(tilemapData.map.copy()) does it still break?

**Comment by matrefeytontias** [comment]
Yep, still remains static.

**Comment by scriptorum** [comment]
Huh, probably because it's a shallow copy. The arrays inside the array are not copied, but still point to the same references.

**Comment by matrefeytontias** [comment]
Yeah, probably. Haxe doesn't have a deep-copy function for arrays, does it ?

**Comment by matrefeytontias** [comment]
Also, I have no idea what's going on:
- Nothing shows up on-screen with the flash build **but** calls to `trace()` still display text normally
- The program instant-crashes, displaying "Uncaught exception - std@module_read" with the Neko build
- Everything works 100% fine with the Windows cpp build (except the whole, animating the tilemap of course)

**Comment by Marc010** [comment]
Can you check if adding ```<haxedef name="haxeJSON" />``` to the project file makes the json array mutable on windows?

**Comment by matrefeytontias** [comment]
Nope, doesn't fix it.

**Comment by matrefeytontias** [comment]
Alright, I figured out the flash part, but the neko part still won't tell me anything about where or why it crashes.

**Comment by matrefeytontias** [comment]
So in the end, just delete the `_map = array;` line and the condition from Tilemap.hx, and that's it ?

**Comment by scriptorum** [comment]
Array does not have a deep copy that I know of but you can 2D loop through the array and .copy() each of those. Something like 
```haxe
var copy = [];
for(row in tilemapData.map)
{
	var newRow = [];
	for(col in arr)
		temp.push(col);
	copy.push(newRow);	
}
```

I'm looking at your update code from [here](http://forum.haxepunk.com/index.php?topic=905.0). For my sanity, what happens when you change the update code to this:

```haxe
    for(y in 0 ... tiles.rows)
    {
        for(x in 0 ... tiles.columns)
        { 
            var curTile = tiles.getTile(x, y);
            if(curTile < 0)
            	continue;

            var mod = curTile % 5;
            var diff = (mod == 4 ? -4 : 1);
			tiles.setTile(x, y, curTile + diff);
        }
    }
```

**Comment by matrefeytontias** [comment]
That's not what it's supposed to do. What you do here is update the tile if it's over -1, but I must only update it if it's over 4. Is that really what you wanted to do ?

**Comment by scriptorum** [comment]
I'm assuming that your code is alternating frames in a sequence of five and repeating. The above does that, skipping over blank (-1) frames. Are you instead using the tile index mod as some sort of timer?

**Comment by matrefeytontias** [comment]
No actually, I just got your code wrong. I use `HXP.alarm()` as a 0.2-seconds timer.

**Comment by matrefeytontias** [comment]
Actually, the thing is that tiles 0 to 4 are not animated. Only tiles 5 to 9, 10 to 14, 15 to 19 and 20 to 24 are. That's why the tiles 0 to 4 must not be cycled through, because they have nothing in common and are not part of the same animation.

**Comment by scriptorum** [comment]
Interesting. With my code I suppose you could continue `if (curTile < 0 || tiles.getIndex(x,y) < 5)`, I just want to verify that your update code isn't part of the problem. As far as I'm concerned, doing a setTile to a random index would also be sufficient to satisfy me if you get the same failure to update. By the way, have you tried taking a look at [TiledSpritemap](https://github.com/HaxePunk/HaxePunk/blob/dev/com/haxepunk/graphics/TiledSpritemap.hx)?

**Comment by matrefeytontias** [comment]
Yeah, the updating part isn't the problem, since even clearing the whole tilemap or just setting a single tile has absolutely no effect not only on the graphics, but also on the _map variable itself - not a single cell changes.
And yes I did take a look, but not all of my animated tiles will be arranged in large and convenient squares. Also, it anti-aliases like hell, and it makes a rather huge contrast with the rest of the tilemap, so I dropped Spritemaps as a whole for the present solution, which works much, much better if you forget the whole immutability problem.

**Comment by scriptorum** [comment]
I will update Tilemap to eliminate this problem. That's really odd behavior.

**Comment by matrefeytontias** [comment]
I think that's really just haxe.Json returning immutable objects, not really HaxePunk's fault.

**Comment by scriptorum** [comment]
I verified that if you do a deep copy of the Json output, it works. What's weird is you CAN change the object directly. When I do `tilemapData.map[3][3] = -1;` it updates the array, but if you make a reference to the map and alter that, *then* it's immutable. So this fails:
```haxe
var map:Array<Array<Int>> = tilemapData.map;
map[3][3] = -1;
```

I feel better that I can see there has been discussion on this over at the Haxe github, notably in issues  https://github.com/HaxeFoundation/haxe/issues/2103 and https://github.com/HaxeFoundation/haxe/issues/2579. It looks like this may be addressed in Haxe 3.2. Apparently casting from Array<Dynamic> to a non-dynamic type causes an implicit copy! Derp.

**Comment by matrefeytontias** [comment]
Alright, so what I'm doing for now is remove the lines we keep talking about in Tilemap.hx. I'll close this now, since it's being taken care of.
